### PR TITLE
[Console] Centralize input stream in base Input class

### PR DESCRIPTION
**Pull request by chalasr**
| Q | A |
| --- | --- |
| Branch? | master |
| Bug fix? | no |
| New feature? | yes |
| BC breaks? | no |
| Deprecations? | yes |
| Tests pass? | yes |
| Fixed tickets | #10844 |
| License | MIT |
| Doc PR | not yet |

Actually we have two classes that have an `inputStream` property with its getter+setter: 
the QuestionHelper ('question' helper) and the SymfonyQuestionHelper (used via SymfonyStyle, that inherits the inputStream from the QuestionHelper class).

This situation makes command testing really hard. 
ATM we can't test a command that uses `SymfonyStyle::ask` without ugly hacks, and we can't find a generic way to set the inputs passed to a command from the CommandTester, as it need to be done on a specific helper that may not be the one used by the command (SymfonyStyle case).

What I propose here is to add a `stream` property (and its getter+setter) to the abstract `Symfony\Component\Console\Input\Input` class.

For now I just made the two helpers setting their `inputStream` from `Input::getStream`, as both QuestionHelper and SymfonyQuestionHelper classes have an `ask` method that takes the command Input as argument.
In a next time (4.0), we could remove the `getInputStream` and `setInputStream` methods from the QuestionHelper class (this only deprecates them in favor of Input `setStream` and `getStream`).

This would close PR #18902 (trying to make interactive command testing with SymfonyStyle easier).
This would also make PR #18710 widely better by setting the input stream in a generic way (working for both helpers without caring about if they are used and which one is used).

Please give me your opinions.

### CODE REVIEW

File: src/Symfony/Component/Console/Input/InputInterface.php [hunk 1]
```diff
@@ -151,4 +151,6 @@ public function isInteractive();
      * @param bool $interactive If the input should be interactive
      */
     public function setInteractive($interactive);
+
+    public function getStream();
```
Comment (ogizanagi) [inline src/Symfony/Component/Console/Input/InputInterface.php]: To revert

Comment (chalasr) [inline src/Symfony/Component/Console/Input/InputInterface.php]: Thank's @ogizanagi, and thank's for having the idea of this.

**Comment by ogizanagi** [comment]
Implementation looks good to me & conform to what we talked about 👍

File: src/Symfony/Component/Console/Application.php [hunk 2]
```diff
@@ -769,8 +769,8 @@ protected function configureIO(InputInterface $input, OutputInterface $output)
 
         if (true === $input->hasParameterOption(array('--no-interaction', '-n'), true)) {
             $input->setInteractive(false);
-        } elseif (function_exists('posix_isatty') && $this->getHelperSet()->has('question')) {
-            $inputStream = $this->getHelperSet()->get('question')->getInputStream();
+        } elseif (function_exists('posix_isatty')) {
+            $inputStream = $input->getStream();
```
Comment (ogizanagi) [inline src/Symfony/Component/Console/Application.php]: Missing `instanceof StreamableInputInterface` check.

File: src/Symfony/Component/Console/Application.php [hunk 3]
```diff
@@ -769,8 +769,8 @@ protected function configureIO(InputInterface $input, OutputInterface $output)
 
         if (true === $input->hasParameterOption(array('--no-interaction', '-n'), true)) {
             $input->setInteractive(false);
-        } elseif (function_exists('posix_isatty') && $this->getHelperSet()->has('question')) {
-            $inputStream = $this->getHelperSet()->get('question')->getInputStream();
```
Comment (ogizanagi) [inline src/Symfony/Component/Console/Application.php]: This should probably not be removed entirely yet ?
Or should `QuestionHelper::setInputStream` && `QuestionHelper::getInputStream` methods call `$input->setStream` && `$input->getStream` ?

**Comment by ogizanagi** [comment]
I think (but nor sure of the policy for such cases) the old tests should be kepts but marked as `@legacy` and new one created. Then, legacy ones will be removed during the merge of this branch in upper branches. (3.3)

File: src/Symfony/Component/Console/Application.php (hunk 3 above)
Comment (chalasr) [inline src/Symfony/Component/Console/Application.php]: Good catch. 
Detecting a possible null helper then null inputStream here would be much logic.
I would prefer calling `Input::setStream` in `QuestionHelper::setInputStream` while setting the `QuestionHelper::$inputStream` (for BC reasons too) then return it in `QuestionHelper::getInputStream()` assuming that if it is set in the QuestionHelper, it will not be from the Input in the meantime.

**Comment by chalasr** [comment]
Re added old tests as legacy.

File: src/Symfony/Component/Console/Helper/QuestionHelper.php [hunk 4]
```diff
@@ -84,6 +91,8 @@ public function setInputStream($stream)
     /**
      * Returns the helper's input stream.
      *
+     * @deprecated Will be removed in 4.0, use StreamableInputInterface::getStream() instead
```
Comment (ogizanagi) [inline src/Symfony/Component/Console/Helper/QuestionHelper.php]: Why did you remove the deprecation triggering ?

File: src/Symfony/Component/Console/Helper/QuestionHelper.php [hunk 5]
```diff
@@ -74,6 +79,8 @@ public function ask(InputInterface $input, OutputInterface $output, Question $qu
      */
     public function setInputStream($stream)
     {
+        @trigger_error(sprintf('The setInputStream() method is deprecated since version 3.1 and will be removed in 4.0. Use %s:setStream() instead.', StreamableInputInterface::class), E_USER_DEPRECATED);
```
Comment (ogizanagi) [inline src/Symfony/Component/Console/Helper/QuestionHelper.php]: Missing `@deprecated` in docblock

Comment (ogizanagi) [inline src/Symfony/Component/Console/Helper/QuestionHelper.php]: You can also use the `__METHOD__` magic constant instead of `setInputStream` to be consistent with other deprecation triggers.

File: src/Symfony/Component/Console/Helper/QuestionHelper.php (hunk 4 above)
Comment (chalasr) [inline src/Symfony/Component/Console/Helper/QuestionHelper.php]: Because we have a call to this getter in Application, so lots of tests as deprecations and don't pass anymore.

File: src/Symfony/Component/Console/Application.php [hunk 6]
```diff
@@ -769,8 +770,18 @@ protected function configureIO(InputInterface $input, OutputInterface $output)
...
+        } elseif (function_exists('posix_isatty')) {
+            $inputStream = null;
+            // Ensure that calling QuestionHelper::setInputStream will works.
+            // To be removed in the same time as getInputStream/setInputStream methods
+            if ($this->getHelperSet()->has('question')) {
+                $inputStream = $this->getHelperSet()->get('question')->getInputStream() ?: $inputStream;
```
Comment (chalasr) [inline src/Symfony/Component/Console/Application.php]: The ternary should be removed

File: src/Symfony/Component/Console/Helper/QuestionHelper.php (hunk 4 above)
Comment (fabpot) [inline src/Symfony/Component/Console/Helper/QuestionHelper.php]: That does not work like that. You MUST add the deprecation notice and fix the tests accordingly. But instead of doing that, I would recommend to just keep `getInputStream()` and add `setInputStream()`.

File: src/Symfony/Component/Console/Tests/Helper/QuestionHelperTest.php [hunk 7]
```diff
@@ -392,6 +385,356 @@ public function testChoiceOutputFormattingQuestionForUtf8Keys()
...
+
+        $question = new ChoiceQuestion('Please select the environment to load', $possibleChoices);
+        $question->setMaxAttempts(1);
+
+        $dialog->ask($this->createInputInterfaceMock(), $this->createOutputInterface(), $question);
+    }
+
+    /**
+     * @requires function mb_strwidth
+     * @legacy
```
Comment (chalasr) [inline src/Symfony/Component/Console/Tests/Helper/QuestionHelperTest.php]: Typo

File: src/Symfony/Component/Console/Tests/Input/InputTest.php [hunk 8]
```diff
@@ -129,4 +129,12 @@ public function testSetGetInteractive()
         $input->setInteractive(false);
         $this->assertFalse($input->isInteractive(), '->setInteractive() changes the interactive flag');
     }
+
+    public function testSetGetStream()
+    {
+        $input = new ArrayInput(array());
+        $stream = fopen('php://memory', 'r+', false);
```
Comment (chalasr) [inline src/Symfony/Component/Console/Tests/Input/InputTest.php]: No need of a resource as there is no check for.

File: src/Symfony/Component/Console/Helper/QuestionHelper.php (hunk 4 above)
Comment (chalasr) [inline src/Symfony/Component/Console/Helper/QuestionHelper.php]: > just keep getInputStream() and add setInputStream()

Does that mean I should only depreciate `setInputStream()` and keep `getInputStream()` as is?

File: src/Symfony/Component/Console/Tests/Input/InputTest.php (hunk 8 above)
Comment (stof) [inline src/Symfony/Component/Console/Tests/Input/InputTest.php]: even if there is no explicit checks, tests should respect the contract of the class.

Comment (chalasr) [inline src/Symfony/Component/Console/Tests/Input/InputTest.php]: 👍

File: src/Symfony/Component/Console/Helper/QuestionHelper.php (hunk 4 above)
Comment (fabpot) [inline src/Symfony/Component/Console/Helper/QuestionHelper.php]: I was more talking about renaming the method in the interface to getInputStream and setInputStream, would that work?

Comment (chalasr) [inline src/Symfony/Component/Console/Helper/QuestionHelper.php]: Not sure that we are talking about the same thing, my description is surely not clear.
 I introduced a new `StreamableInputInterface` implemented by `Symfony\Component\Console\Input\Input` containing two methods `getStream()` and  `setStream()`.
These methods would replace the old `QuestionHelper::getInputStream()` and `QuestionHelper::setInputStream()`, that's why I depreciate them here.
The problem is that `Application::configureIO` get the stream from `QuestionHelper::getInputStream()` before, so we need to keep it intact before totally remove it later. But, this call causes deprecations in all ApplicationTest tests that call `Application::run()` because of the call of the deprecated `QuestionHelper::getInputStream()`.

**Comment by stof** [comment]
@chalasr would be better to avoid moving them, but only marking them as legacy, so that we have less conflicts when merging branches

**Comment by chalasr** [comment]
@stof Reverted.

**Comment by fabpot** [comment]
`getInputStream()` methods should also be deprecated, right?

**Comment by fabpot** [comment]
In the description, you mention `SymfonyQuestionHelper`, but this class is not changed in this PR.

**Comment by chalasr** [comment]
@fabpot Yes the SymfonyQuestionHelper class is not changed, it would just become testable via the CommandTester as well as the QuestionHelper.

About the `getInputStream()`  method you're right, it should also be deprecated. 
The problem is that the method [is called](https://github.com/symfony/symfony/pull/18999/files#diff-38eebdda7d6ebfcb7405b91167f05c9cR783) in ``Application::configureIO` to preserve BC, that causes lots of deprecation warnings from ApplicationTest, each time `Application::run()` is called.

As there is no reason to call `getInputStream()` without `setInputStream()`, I assumed that depreciate the `setInputStream()` method would be sufficient.
Seeing your comment I guess that doesn't work like that, right? :)

What do you think about this?

**Comment by fabpot** [comment]
@chalasr Here is how you can make it work: add a `$triggerDeprecationError = true` argument to `getInputStream()` and only trigger the deprecation when `$triggerDeprecationError` is true. When you call it yourself, pass `false`.

**Comment by chalasr** [comment]
The `getInputStream()` method is now deprecated.
@fabpot Thank you!

**Comment by fabpot** [comment]
Thank you @chalasr.

File: src/Symfony/Component/Console/Helper/QuestionHelper.php [hunk 9]
```diff
@@ -84,10 +94,17 @@ public function setInputStream($stream)
     /**
      * Returns the helper's input stream.
      *
+     * @deprecated since version 3.2, to be removed in 4.0. Use
+     *             StreamableInputInterface::getStream() instead.
+     *
      * @return resource
      */
-    public function getInputStream()
+    public function getInputStream($triggerDeprecationError = true)
```
Comment (xabbuh) [inline src/Symfony/Component/Console/Helper/QuestionHelper.php]: adding a new argument here is a BC break

Comment (xabbuh) [inline src/Symfony/Component/Console/Helper/QuestionHelper.php]: see https://github.com/symfony/symfony/pull/19072

**Comment by nicolas-grekas** [comment]
@chalasr can you please check the remaining deprecated call on master and verify that all usage of the deprecated method have been removed from master (except when testing it for legacy reasons)?

**Comment by chalasr** [comment]
@nicolas-grekas Just verified, there is no more calls to these methods in master, except for tests marked as legacy in `Symfony\Component\Console\Tests\Helper\QuestionHelperTest` and `Symfony\Component\Console\Tests\Helper\SymfonyQuestionHelperTest`, and (of course), the last call of `getInputStream()` ensuring BC in `Symfony\Component\Console\Application`.
